stl/inc/ranges

```diff
@@ -33,10 +33,195 @@ namespace ranges {
+                _Pipeline{static_cast<_Derived&&>(__l), static_cast<const _Other&>(__r)})) {
+                // clang-format on
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Derived, _Base<_Derived>>);
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Other, _Base<_Other>>);
+                return _Pipeline{static_cast<_Derived&&>(__l), static_cast<const _Other&>(__r)};
+            }
+
+            // clang-format off
+            template <class _Other>
+                requires _Can_compose<const _Derived&, _Other&&>
+            friend constexpr auto operator|(const _Base& __l, _Base<_Other>&& __r) noexcept(noexcept(
+                _Pipeline{static_cast<const _Derived&>(__l), static_cast<_Other&&>(__r)})) {
+                // clang-format on
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Derived, _Base<_Derived>>);
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Other, _Base<_Other>>);
+                return _Pipeline{static_cast<const _Derived&>(__l), static_cast<_Other&&>(__r)};
+            }
+
+            // clang-format off
+            template <class _Other>
+                requires _Can_compose<const _Derived&, const _Other&>
+            friend constexpr auto operator|(const _Base& __l, const _Base<_Other>& __r) noexcept(noexcept(
+                _Pipeline{static_cast<const _Derived&>(__l), static_cast<const _Other&>(__r)})) {
+                // clang-format on
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Derived, _Base<_Derived>>);
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Other, _Base<_Other>>);
+                return _Pipeline{static_cast<const _Derived&>(__l), static_cast<const _Other&>(__r)};
+            }
+
+            template <_Can_pipe<const _Derived&> _Left>
+            friend constexpr auto operator|(_Left&& __l, const _Base& __r) noexcept(
+                noexcept(static_cast<const _Derived&>(__r)(static_cast<_Left&&>(__l)))) {
+                return static_cast<const _Derived&>(__r)(static_cast<_Left&&>(__l));
+            }
+
+            template <_Can_pipe<_Derived> _Left>
+            friend constexpr auto operator|(_Left&& __l, _Base&& __r) noexcept(
+                noexcept(static_cast<_Derived&&>(__r)(static_cast<_Left&&>(__l)))) {
+                return static_cast<_Derived&&>(__r)(static_cast<_Left&&>(__l));
+            }
+        };
+
+        template <class _Left, class _Right>
+        struct _Pipeline : _Base<_Pipeline<_Left, _Right>> {
+            /* [[no_unique_address]] */ _Left __l;
+            /* [[no_unique_address]] */ _Right __r;
+
+            template <class _Ty1, class _Ty2>
+            constexpr explicit _Pipeline(_Ty1&& _Val1, _Ty2&& _Val2) noexcept(
+                is_nothrow_convertible_v<_Ty1, _Left>&& is_nothrow_convertible_v<_Ty2, _Right>)
+                : __l(static_cast<_Ty1&&>(_Val1)), __r(static_cast<_Ty2&&>(_Val2)) {}

```

Comment:
Is there a reason that these are all phrased with `static_cast` instead of `forward`?

Reply:
These two should unequivocally be `forward`s. I missed them arguing with myself over the `operator()`s: is it better to use `forward` in the cases that occur after overload resolution, or to consistently use `static_cast` so it's obvious that equivalent expressions are identical? I think I've been waffling on that usage from PR to PR. A macro as proposed in #1208 would resolve this. For the time being I'll try to consistently use `static_cast` "before" overload resolution and `forward` after; hopefully appearances of both in close proximity will help to remind folks that they're equivalent.